lgc/builder/DescBuilder.cpp

```diff
@@ -411,55 +430,66 @@ Value *DescBuilder::getDescPtr(ResourceNodeType resType, unsigned descSet, unsig
-  if (node && node == topNode) {
+  static unsigned i = 0;
+  llvm::errs() << "getDescPtr " << ++i << ", resType: " << resType << ", node: " << node << "\n";
+
+  auto GetSpillTableAddr = [this]() -> Value * {
     // The descriptor is in the top-level table. (This can only happen for a DescriptorBuffer.) Contrary
     // to what used to happen, we just load from the spill table, so we can get a pointer to the descriptor.
     // The spill table gets returned as a pointer to array of i8.
-    descPtr =
-        CreateNamedCall(lgcName::SpillTable, getInt8Ty()->getPointerTo(ADDR_SPACE_CONST), {}, Attribute::ReadNone);
-    // Ensure we mark spill table usage.
-    getPipelineState()->getPalMetadata()->setUserDataSpillUsage(node->offsetInDwords);
-  } else {
+    return CreateNamedCall(lgcName::SpillTable, getInt8Ty()->getPointerTo(ADDR_SPACE_CONST), {}, Attribute::ReadNone);
+  };
+
+  auto GetDescriptorSetAddr = [this, descSet, shadow]() -> Value * {
     // Get the descriptor table pointer for the set, which might be passed as a user SGPR to the shader.
     // The args to the lgc.descriptor.set call are:
     // - descriptor set number
     // - value for high 32 bits of pointer; HighAddrPc to use PC
     // TODO Shader compilation: For the "shadow" case, the high half of the address needs to be a reloc.
     unsigned highHalf = shadow ? m_pipelineState->getOptions().shadowDescriptorTable : HighAddrPc;
-    descPtr = CreateNamedCall(lgcName::DescriptorSet, getInt8Ty()->getPointerTo(ADDR_SPACE_CONST),
-                              {getInt32(descSet), getInt32(highHalf)}, Attribute::ReadNone);
+    return CreateNamedCall(lgcName::DescriptorSet, getInt8Ty()->getPointerTo(ADDR_SPACE_CONST),
+                           {getInt32(descSet), getInt32(highHalf)}, Attribute::ReadNone);
+  };
+
+  // Get the descriptor table pointer.
+  if (node && node == topNode) {
+    // Ensure we mark spill table usage.
+    descPtr = GetSpillTableAddr();
+    getPipelineState()->getPalMetadata()->setUserDataSpillUsage(node->offsetInDwords);
+    llvm::errs() << "\tcase 1) topNode, offset " << node->offsetInDwords << "\n";
+  } else if (!node && !topNode && resType == ResourceNodeType::DescriptorBuffer) {
+    // If we do not have user data layout info (topNode and node are nullptr), then
+    // we do not know at compile time whether a DescriptorBuffer is in the root table or the table for its
+    // descriptor set, so we need to generate a select between the two, where the condition is a reloc.
+    // If the descriptor ends up in the root table (top-level), a value from the spill table will be used.
+    // The linking code has to take care of marking PAL metadata for user spill usage.
+    Value *spillDescPtr = GetSpillTableAddr();
+    Value *descriptorTableDescPtr = GetDescriptorSetAddr();
+    Value *reloc = CreateRelocationConstant(reloc::DescriptorUseSpillTable + Twine(descSet) + "_" + Twine(binding) +
```

Comment:
I think we don't need a suffix here, because a descriptor in the root table can only be a buffer. My understanding is that it represents a Vulkan "dynamic descriptor", which is a buffer where you can change the offset without needing a new PSO.